### PR TITLE
Switch study public/private

### PR DESCRIPTION
**Pull request by AbdelHedhili**

### CODE REVIEW

File: src/main/java/org/gridsuite/study/server/StudyService.java [hunk 1]
```diff
@@ -592,6 +593,36 @@ private LoadFlowResult jsonToLoadFlowResult(String strLfResult) {
...
+        Mono<StudyEntity> studyEntityMono = getStudy(studyName, userId);
+        return studyEntityMono.switchIfEmpty(Mono.error(new StudyException(STUDY_NOT_FOUND))).flatMap(studyEntity -> {
+            Mono<Void> delete;
+            Mono<StudyEntity> insert;
+            //if the study is public and we want to make it private
+            if (!studyEntity.isPrivate() && toPrivate) {
+                delete = studyRepository.deleteStudy(userId, studyName);
+                insert = studyRepository.insertStudy(studyEntity.getStudyName(), userId, true, studyEntity.getNetworkUuid(),
+                        studyEntity.getNetworkId(), studyEntity.getDescription(), studyEntity.getCaseFormat(), studyEntity.getCaseUuid(),
+                        studyEntity.isCasePrivate(), new LoadFlowResult(studyEntity.getLoadFlowResult().getStatus()), studyEntity.getLoadFlowParameters());
```
Comment (geofjamg) [inline src/main/java/org/gridsuite/study/server/StudyService.java:611]: We loose security analysis result UUID

File: src/main/java/org/gridsuite/study/server/StudyService.java [hunk 2]
```diff
@@ -592,6 +593,36 @@ private LoadFlowResult jsonToLoadFlowResult(String strLfResult) {
...
+              //if the study is private and we want to make it public
+            } else if (studyEntity.isPrivate() && !toPrivate) {
+                delete = studyRepository.deleteStudy(userId, studyName);
+                insert = studyRepository.insertStudy(studyEntity.getStudyName(), userId, false, studyEntity.getNetworkUuid(),
```
Comment (geofjamg) [inline src/main/java/org/gridsuite/study/server/StudyService.java:615]: Same

File: src/main/java/org/gridsuite/study/server/dto/StudyInfos.java [hunk 3]
```diff
@@ -24,5 +24,7 @@
 
     String caseFormat;
 
+    Boolean isPrivate;
```
Comment (geofjamg) [inline src/main/java/org/gridsuite/study/server/dto/StudyInfos.java:27]: boolean ?

File: src/test/java/org/gridsuite/study/server/StudyTest.java [hunk 4]
```diff
@@ -777,7 +851,8 @@ public boolean matchesSafely(StudyInfos s) {
             return super.matchesSafely(s)
                     && source.getCaseFormat().equals(s.getCaseFormat())
                     && source.getDescription().equals(s.getDescription())
-                    && source.getLoadFlowResult().getStatus() == s.getLoadFlowResult().getStatus();
+                    && source.getLoadFlowResult().getStatus() == s.getLoadFlowResult().getStatus()
+                    && source.getIsPrivate().equals(s.getIsPrivate());
```
Comment (geofjamg) [inline src/test/java/org/gridsuite/study/server/StudyTest.java:855]: == if boolean?

**Review by geofjamg — CHANGES_REQUESTED** [review]

File: src/main/java/org/gridsuite/study/server/StudyService.java [hunk 5]
```diff
@@ -592,6 +593,36 @@ private LoadFlowResult jsonToLoadFlowResult(String strLfResult) {
...
+                        studyEntity.getNetworkId(), studyEntity.getDescription(), studyEntity.getCaseFormat(), studyEntity.getCaseUuid(),
+                        studyEntity.isCasePrivate(), new LoadFlowResult(studyEntity.getLoadFlowResult().getStatus()), studyEntity.getLoadFlowParameters());
+            } else {
+                return Mono.just(studyEntity);
+            }
+            return delete.then(insert);
+        }).map(StudyService::toInfos);
+
```
Comment (SlimaneAmar) [inline src/main/java/org/gridsuite/study/server/StudyService.java:623]: Maybe a more concise form?
         return getStudy (studyName, userId) .switchIfEmpty (Mono.error (new StudyException (STUDY_NOT_FOUND))). flatMap (studyEntity ->
             (studyEntity.isPrivate () == toPrivate)? Mono.just (studyEntity):
                     studyRepository.deleteStudy (userId, studyName)
                     .then (insertStudy (studyEntity.getStudyName (), userId, toPrivate, studyEntity.getNetworkUuid (),
                             studyEntity.getNetworkId (), studyEntity.getDescription (), studyEntity.getCaseFormat (), studyEntity.getCaseUuid (),
                             studyEntity.isCasePrivate (), new LoadFlowResult (studyEntity.getLoadFlowResult (). getStatus ()), studyEntity.getLoadFlowParameters ()))
         ) .map (StudyService :: toInfos);

File: src/test/java/org/gridsuite/study/server/StudyTest.java [hunk 6]
```diff
@@ -698,6 +717,61 @@ public MockResponse dispatch(RecordedRequest request) {
...
+                        .creationDate(ZonedDateTime.now(ZoneId.of("UTC")))
+                        .loadFlowResult(new LoadFlowResult(LoadFlowStatus.CONVERGED)).build()));
+
+        // try to change access right for a study of another user -> forbidden
+        webTestClient.post()
+                .uri("/v1/userId/studies/{studyName}/accessRights?toPrivate=true", newStudyName)
+                .header("userId", "notAuth")
+                .exchange()
+                .expectStatus().isForbidden();
+
```
Comment (SlimaneAmar) [inline src/test/java/org/gridsuite/study/server/StudyTest.java:774]: Missing the non-existent study case test ?

**Review by SlimaneAmar — APPROVED** [review]

**Review by geofjamg — APPROVED** [review]

**Pull request merged**